### ISSUE TITLE
"12 bytes lost due to alignment ..." error on magic_wand example for SparkFun Edge

### ISSUE THREAD

**Issue by pfliegster** [tensorflow/tensorflow#40837]
@tensorflow/micro

**System information**
- Host OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac OSX Catalina (10.15.5)
- TensorFlow installed from (source or binary): source
- Tensorflow version (commit SHA if source): 44f2f014e5d84e195af7de6f9d9400df3d3a00dd
- Target platform (e.g. Arm Mbed OS, Arduino Nano 33 etc.): SparkFun Edge

**Describe the problem**
Cannot view/display any debug information from UART output but receive an error instead, using screen command: 

`screen ${DEVICENAME} 115200 `

I also tried to explicitly substitute the actual device name, but I get the same error result. I only get this single line of output, indicating something is wrong:

`12 bytes lost due to alignment. To avoid this loss, please make sure the tensor_arena is 16 bytes aligned.`

Note: I have followed this same sequence (and settings) on other examples for SparkFun Edge (**Hello World** and **Person Detection**) and for both of those builds, I do get proper/expected UART debug output using screen, so this error seems specific to the magic_wand example.

Every time I then reset the board, I get this same error repeated in screen.

**Please provide the exact sequence of commands/steps when you ran into the problem**
I followed the readme instructions for the example exactly:
tensorflow/lite/micro/examples/magic_wand/README.md, under the "Deploy to SparkFun Edge", Specifically, from the root directory cloned from GitHub, I executed these commands:

```
make -f tensorflow/lite/micro/tools/make/Makefile TARGET=sparkfun_edge magic_wand_bin

cp tensorflow/lite/micro/tools/make/downloads/AmbiqSuite-Rel2.2.0/tools/apollo3_scripts/keys_info0.py \
tensorflow/lite/micro/tools/make/downloads/AmbiqSuite-Rel2.2.0/tools/apollo3_scripts/keys_info.py

python3 tensorflow/lite/micro/tools/make/downloads/AmbiqSuite-Rel2.2.0/tools/apollo3_scripts/create_cust_image_blob.py \
--bin tensorflow/lite/micro/tools/make/gen/sparkfun_edge_cortex-m4/bin/magic_wand.bin \
--load-address 0xC000 \
--magic-num 0xCB \
-o main_nonsecure_ota \
--version 0x0

python3 tensorflow/lite/micro/tools/make/downloads/AmbiqSuite-Rel2.2.0/tools/apollo3_scripts/create_cust_wireupdate_blob.py \
--load-address 0x20000 \
--bin main_nonsecure_ota.bin \
-i 6 \
-o main_nonsecure_wire \
--options 0x1

export BAUD_RATE=921600
export DEVICENAME=/dev/tty.usbserial-A50285BI
```
and then flash the binary using the boatload button sequence as described while executing this command:

```
python3 tensorflow/lite/micro/tools/make/downloads/AmbiqSuite-Rel2.2.0/tools/apollo3_scripts/uart_wired_update.py \
-b ${BAUD_RATE} ${DEVICENAME} \
-r 1 \
-f main_nonsecure_wire.bin \
-i 6
```

All commands seem to execute correctly, no errors displayed, and boatload process is successful. After hitting the board reset button, I get the yellow light blinking rapidly after a few seconds (I assume it is working at this point). Then, using the screen command to view UART output, I get the above error instead of the expected output of "Magic Starts!".

**Comment by lheim** [comment]
Hi @pfliegster,
Your UART communication is working, this is where you get the error from.

Did you modify `kTensorArenaSize` in the `main_functions.cc`? Try playing around with this.

Otherwise you might try one of the already given Keras or TF Lite models instead of your own: https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/micro/examples/magic_wand/train/netmodels/CNN

**Comment by pfliegster** [comment]
@lahwran - thanks for the suggestion. I'm am quite new to this code base, so I really haven't tried to tweak anything yet, but rather just run the built-in examples as-is.

However, following your train of thought ... looking at the definition for `kTensorArenaSize` in `main_functions.cc` for this example, I'm guessing modifications here will change the size of `tensor_arena[]`, but otherwise not directly affect it's alignment in memory ... if I understand the error correctly, I would think I need some kind of linker script command or pre-processor directive to ensure alignment of this buffer on a 16-byte boundary.

Currently `kTensorArenaSize = 60 * 1024;`, so it this buffer will naturally be a multiple of 16-bytes in size, I just don't know how to control its alignment (or fix whatever the true cause of that error message might be).

So ... just to test, I tried just adding 4, 8, and 12 to the size value, but each of these creates the same erroneous output. I suspect that this example project is just missing some alignment mechanism that is there in some of the other examples, or there is some other more intrinsic problem with this example. I'll try to dig into this a bit on my own, but hoped the TF gardeners who know the code best (or pros like yourself who have experience navigating around this code base) would have a ready solution.

Thanks for the idea & for the reference on training my own model too, I'll look that over.

**Comment by kcoh7777** [comment]
Dear,
Please try to Insert "alignas(16)" in the line that defines tensor_arena in main_functions.cc like below:
  alignas(16) uint8_t tensor_arena[kTensorArenaSize];

**Comment by pfliegster** [comment]
@kcoh7777 - Thank you! That fixed the issue ... you're a genius! I really appreciate the help.

Of course, now it only recognizes "slope", but I think that is a different issue. Anyway, I don't know if exactly how to get the solution/fix pushed back into the git repo (or if I even have privilege to do so), but if you do - please do so. Thanks!
